en/contributing/test.md

```diff
@@ -77,3 +87,138 @@ To run all telemetry tests:
+```cpp
+#include "example_impl.h"
+#include "dronecore.h"
+#include "global_include.h"
+#include <gtest/gtest.h>
+
+namespace dronecore {
+
+TEST(ExampleImpl, NoTest)
+{
+    ASSERT_TRUE(true);
+}
+
+} // namespace dronecore
+```
+
+#### Adding Unit Tests
+
+In order to run a unit test its file needs to be added to the DroneCore unit test program (`unit_tests_runner`). For the core plugins this is done in the main DroneCore [DroneCore/CMakeLists.txt](https://github.com/dronecore/DroneCore/blob/master/CMakeLists.txt#L187) file:
+```cmake
+    add_executable(unit_tests_runner
+        core/global_include_test.cpp
+        core/mavlink_channels_test.cpp
+        core/unittests_main.cpp
+        core/http_loader_test.cpp
+        core/timeout_handler_test.cpp
+        core/call_every_handler_test.cpp
+        ${plugin_unittest_source_files}
+        ${unit_tests_src}
+    )
+```
+
+### Writing Integration Tests
+
+DroneCore provides the `integration_tests_runner` application for running the integration tests and 
+some helper code to make it easier to log tests and run them against the simulator.
+
+> **Tip** Check out the [Google Test Primer](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md) 
+> and the [integration_tests](https://github.com/dronecore/DroneCore/tree/develop/integration_tests) 
+> for our existing plugins to better understand how to write your own!
+
+#### Integration Test Files/Code
+
+The main DroneCore-specific functionality is provided by [integration_test_helper.h](https://github.com/dronecore/DroneCore/blob/master/core/integration_test_helper.h). 
+This provides access to the [Plugin/Test Logger](../contributing/dev_logging.md) 
+and a shared test class `SitlTest` for setting up and tearing down the PX4 simulator.
+
+> **Note** All tests must be declared using `TEST_F` and have a first argument `SitlTest` as shown. This is required
+> in order to use the shared class to set up and tear down the simulator between tests.
+
+The example integration test [hello_world.cpp](https://github.com/dronecore/DroneCore/blob/master/external_example/integration_tests/hello_world.cpp) demonstrates this below. 
+
+```cpp
+#include <iostream>
+#include <unistd.h>
+#include "dronecore.h"
+#include "integration_test_helper.h"
+
+using namespace dronecore;
+
+TEST_F(SitlTest, ExternalExampleHello)
+{
+    DroneCore dc;
+
+    DroneCore::ConnectionResult ret = dc.add_udp_connection();
+    ASSERT_EQ(ret, DroneCore::ConnectionResult::SUCCESS);
+
+    // Wait for device to connect via heartbeat.
+    std::this_thread::sleep_for(std::chrono::seconds(2));
+
+    // One vehicle should have connected.
+    std::vector<uint64_t> uuids = dc.device_uuids();
+    EXPECT_EQ(uuids.size(), 1);
+
+    // Apparently it can say hello.
+    dc.device().example().say_hello();
+}
+```
+
+#### Adding Integration Tests
+
```

Comment:
@julianoes I think this can be simplifed in that looking at this external and internal plugins do the same thing to add integration tests. It is just that the CMakeLists.txt is in different places. So I'd tidy this to make that clear.